File: Leetcode/C++/300. Longest Increasing Subsequence/solution.cpp
// https://leetcode.com/problems/longest-increasing-subsequence/description/

#include <algorithm>
#include <vector>

class Solution
{
public:
    int lengthOfLIS(std::vector<int> &nums)
    {
        std::vector<int> dp;
        for (int idx = 0; idx < nums.size(); idx++)
        {
            dp.push_back(1);
            for (int i = 0; i < idx; i++)
            {
                if (nums[i] < nums[idx])
                {
                    dp[idx] = std::max(dp[idx], dp[i] + 1);
                }
            }
        }
        return *max_element(dp.begin(), dp.end());
    }
};

File: Leetcode/Python/300. Longest Increasing Subsequence/solution.py
# https://leetcode.com/problems/longest-increasing-subsequence/description/
# Time complexity: O(n log n)

from bisect import bisect_left


class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        lis = []
        for num in nums:
            if not lis or lis[-1] < num:
                lis.append(num)
                continue
            lis[bisect_left(lis, num)] = num
        return len(lis)
